File: osx/Sources/AutoSwitchKVM/USB/SourceLearner.swift
import Foundation
import Combine

/// Detects the USB device(s) that constitute a KVM source by watching what changes while the
/// user switches the KVM. Snapshot the attached devices at start; any device that appears or
/// disappears during the window becomes a candidate the user then reviews and names.
@MainActor
final class SourceLearner: ObservableObject {
    @Published private(set) var isLearning = false
    @Published private(set) var changeCount = 0

    private let usb: USBMonitoring
    private var cancellable: AnyCancellable?
    private var baseline: [UInt32: USBDeviceInfo] = [:]
    private var changedKeys = Set<UInt32>()

    init(usb: USBMonitoring) { self.usb = usb }

    private func key(_ vendorID: UInt16, _ productID: UInt16) -> UInt32 {
        (UInt32(vendorID) << 16) | UInt32(productID)
    }
    private func key(_ d: USBDeviceInfo) -> UInt32 { key(d.vendorID, d.productID) }

    func start() {
        baseline = Dictionary(usb.attachedDevices.map { (key($0), $0) }, uniquingKeysWith: { a, _ in a })
        changedKeys = []
        changeCount = 0
        isLearning = true
        cancellable = usb.events.sink { [weak self] event in
            guard let self else { return }
            self.changedKeys.insert(self.key(event.vendorID, event.productID))
            self.changeCount = self.changedKeys.count
        }
    }

    /// Stop learning and return the candidate devices (appeared or disappeared during the window),
    /// with names resolved from the current attached list or the start-of-window snapshot.
    func finish() -> [USBDeviceInfo] {
        isLearning = false
        cancellable?.cancel(); cancellable = nil

        let current = Dictionary(usb.attachedDevices.map { (key($0), $0) }, uniquingKeysWith: { a, _ in a })
        var result: [USBDeviceInfo] = []
        for k in changedKeys {
            if let info = current[k] ?? baseline[k] {
                result.append(info)
            } else {
                result.append(USBDeviceInfo(vendorID: UInt16(k >> 16),
                                            productID: UInt16(k & 0xFFFF),
                                            name: ""))
            }
        }
        return result.sorted { $0.displayName < $1.displayName }
    }

    func cancel() {
        isLearning = false
        cancellable?.cancel(); cancellable = nil
        changedKeys = []
        changeCount = 0
    }
}

File: windows/src/AutoSwitchKVM.Core/SourceLearner.cs
namespace AutoSwitchKVM.Core;

/// Detects the USB device(s) that make up a KVM source by watching what changes while the user
/// switches the KVM. Ported from osx/.../USB/SourceLearner.swift, adapted to the snapshot-based
/// IUsbMonitor: snapshot the attached devices at Start; any (vendor,product) that appears or
/// disappears vs that baseline during the window becomes a candidate to review and name.
public sealed class SourceLearner
{
    private readonly IUsbMonitor _usb;
    private Dictionary<uint, UsbDeviceInfo> _baseline = new();
    private readonly HashSet<uint> _changedKeys = new();
    private Action<IReadOnlyList<UsbDeviceInfo>>? _handler;

    public bool IsLearning { get; private set; }
    public int ChangeCount => _changedKeys.Count;

    /// Fires when ChangeCount updates (so the UI can show progress).
    public event Action? Changed;

    public SourceLearner(IUsbMonitor usb) => _usb = usb;

    private static uint Key(ushort vendor, ushort product) => ((uint)vendor << 16) | product;
    private static uint Key(UsbDeviceInfo d) => Key(d.VendorID, d.ProductID);

    public void Start()
    {
        _baseline = Snapshot();
        _changedKeys.Clear();
        IsLearning = true;
        _handler = OnSnapshot;
        _usb.Changed += _handler;
        Changed?.Invoke();
    }

    private void OnSnapshot(IReadOnlyList<UsbDeviceInfo> snapshot)
    {
        var currentKeys = snapshot.Select(Key).ToHashSet();
        var baselineKeys = _baseline.Keys.ToHashSet();
        foreach (var k in currentKeys.Except(baselineKeys)) _changedKeys.Add(k);   // appeared
        foreach (var k in baselineKeys.Except(currentKeys)) _changedKeys.Add(k);   // disappeared
        Changed?.Invoke();
    }

    /// Stop learning and return the candidate devices (appeared or disappeared during the window),
    /// names resolved from the current attached list or the start-of-window snapshot.
    public IReadOnlyList<UsbDeviceInfo> Finish()
    {
        IsLearning = false;
        Unsubscribe();

        var current = Snapshot();
        var result = new List<UsbDeviceInfo>();
        foreach (var k in _changedKeys)
        {
            if (current.TryGetValue(k, out var info) || _baseline.TryGetValue(k, out info))
                result.Add(info);
            else
                result.Add(new UsbDeviceInfo((ushort)(k >> 16), (ushort)(k & 0xFFFF)));
        }
        return result.OrderBy(d => d.DisplayName, StringComparer.Ordinal).ToList();
    }

    public void Cancel()
    {
        IsLearning = false;
        Unsubscribe();
        _changedKeys.Clear();
        Changed?.Invoke();
    }

    private Dictionary<uint, UsbDeviceInfo> Snapshot()
    {
        var map = new Dictionary<uint, UsbDeviceInfo>();
        foreach (var d in _usb.Snapshot())
            map[Key(d)] = d;   // last writer wins (de-dup by vid/pid)
        return map;
    }

    private void Unsubscribe()
    {
        if (_handler != null) { _usb.Changed -= _handler; _handler = null; }
    }
}
